process/smartContract/builtInFunctions/keyValueStorage.go

```diff
@@ -37,58 +37,76 @@ func (k *saveKeyValueStorage) ProcessBuiltinFunction(
 	_, acntDst state.UserAccountHandler,
 	input *vmcommon.ContractCallInput,
 ) (*vmcommon.VMOutput, error) {
+	err := checkArgumentsForSaveKeyValue(acntDst, input)
+	if err != nil {
+		return nil, err
+	}
+
+	vmOutput := &vmcommon.VMOutput{
+		GasRemaining: input.GasProvided,
+	}
+
+	useGas := k.funcGasCost
+	for i := 0; i < len(input.Arguments); i += 2 {
+		key := input.Arguments[i]
+		value := input.Arguments[i+1]
+		length := uint64(len(value))
+		useGas += length * (k.gasConfig.DataCopyPerByte + k.gasConfig.PersistPerByte)
+
+		if !process.IsAllowedToSaveUnderKey(key) {
+			return nil, fmt.Errorf("%w it is not allowed to save under key %s", process.ErrOperationNotPermitted, key)
+		}
+
+		oldValue, _ := acntDst.DataTrieTracker().RetrieveValue(key)
+		if bytes.Equal(oldValue, value) {
+			continue
+		}
+
+		lengthChange := uint64(0)
+		lengthOldValue := uint64(len(oldValue))
+		if lengthOldValue < length {
+			lengthChange = length - lengthOldValue
+		} else {
+			releaseLength := lengthOldValue - length
+			vmOutput.GasRefund = big.NewInt(0).Mul(big.NewInt(0).SetUint64(releaseLength), big.NewInt(0).SetUint64(k.gasConfig.ReleasePerByte))
+		}
+
+		useGas += k.gasConfig.StorePerByte * lengthChange
+		if input.GasProvided < useGas {
+			return nil, process.ErrNotEnoughGas
+		}
+
+		acntDst.DataTrieTracker().SaveKeyValue(key, value)
```

Comment:
This means that we don't set key-values atomically - is this what we want? Perhaps a rollback is actually done in case of `ErrNotEnoughGas` after setting, say, 2 key-values, but not having enough gas for the remaining, say, 5 key-values?

Reply:
if error is done - rollback is processed for all save key values. you can either save all, or none at all.